src/matrix.hpp

```diff
@@ -0,0 +1,126 @@
+
+    // Defaulted move constructors are auto noexcept
+    // NOLINTBEGIN(performance-noexcept-move-constructor)
+
+    matrix() = default;
+    ~matrix() = default;
+    matrix(matrix&&) = default;
+    matrix(const matrix&) = default;
+    auto operator=(matrix&&) -> matrix& = default;
+    auto operator=(const matrix&) -> matrix& = default;
+
+    // NOLINTEND(performance-noexcept-move-constructor)
+
+    template <class... Args>  // clang-format off
+        requires(
+            (std::same_as<columns_type, std::remove_cvref_t<Args>> && ...) &&
+            (sizeof...(Args) == cols))
+    constexpr matrix(Args&&... args)  // clang-format on
+        : columns{std::forward<Args>(args)...}
+    {}
+
+    // TODO: add construction by row
+
+    [[nodiscard]] constexpr auto begin() & noexcept { return columns.begin(); }
+    [[nodiscard]] constexpr auto begin() const& noexcept
+    {
+        return columns.begin();
+    }
+    [[nodiscard]] constexpr auto cbegin() const& noexcept
+    {
+        return columns.cbegin();
+    }
+
+    [[nodiscard]] constexpr auto end() & noexcept { return columns.end(); }
+    [[nodiscard]] constexpr auto end() const& noexcept { return columns.end(); }
+    [[nodiscard]] constexpr auto cend() const& noexcept
+    {
+        return columns.cend();
+    }
+
+    [[nodiscard]] constexpr auto
+    operator[](typename columns_type::coords_type::size_type n) & -> auto&
+    {
+        return columns[n];
+    }
+    [[nodiscard]] constexpr auto
+    operator[](typename columns_type::coords_type::size_type n) const& -> auto&
+    {
+        return columns[n];
+    }
+
+    [[nodiscard]] constexpr auto
+    operator[](std::pair<typename columns_type::coords_type::size_type,
+                         typename matrix::size_type> indices) & -> auto&
+    {
+        return columns[indices.first][indices.second];
+    }
+    [[nodiscard]] constexpr auto
+    operator[](std::pair<typename columns_type::coords_type::size_type,
+                         typename matrix::size_type> indices) const& -> auto&
+    {
+        return columns[indices.first][indices.second];
+    }
+
+    friend auto operator<<(std::ostream& os, const matrix& p) -> std::ostream&
+    {
+        os << "[" << std::endl;
+
+        // TODO use iota, or create a matrix view that traverses the matrix by
+        // rows
+        for (matrix::size_type r{0}; r < matrix::rows; ++r) {
+            os << p[{r, 0}];
+            // TODO use `std::views::drop` once clang implements it
+            for (const auto& x : std::ranges::drop_view{p, 1}) {
+                os << ", " << x[r];
+            }
+            os << ";" << std::endl;
+        }
+        os << "]";
+
+        return os;
+    }
+
+    [[nodiscard]] friend constexpr auto
+    operator==(const matrix& lhs, const matrix& rhs) -> bool = default;
+
+    [[nodiscard]] friend constexpr auto
+    operator*(const matrix<T, Rows, Cols>& m, const vector<T, Cols>& v)
+        -> matrix<T, Rows, 1>
```

Comment:
Is this the right return type? Should this be `vector`? Should `vector` be an alias for `matrix<T, R, 1>`?

Reply:
I am going back and forth. But I think they should be different types. I would like to avoid the choices they made with numpy, I don't like the confusion between the two

Reply:
In the end, I removed operator* for vectors. Now only matrices can be multiplied

Reply:
are vectors promoted to matrices automatically?

Reply:
No, and I am not sure they should be since in principle a vector can be used to generate a row or a column matrix. But maybe it is just a matter of using the correct requires (i.e, M x N  * vector<N> is ok (vector promoted to column), M x 1 * vector<N> is also ok (vector promoted to row matrix), and same for left multiplication?

Reply:
A vector should probably be promotable to either row or column? Or what if you just make vector an alias for column vector?